web-api/benches/coi_personalization.rs

```diff
@@ -0,0 +1,162 @@
+    rngs::SmallRng,
+    SeedableRng,
+};
+use xayn_ai_bert::Embedding1;
+use xayn_ai_coi::{CoiConfig, CoiId, CoiStats, PositiveCoi, UserInterests};
+use xayn_web_api::bench_rerank;
+
+macro_rules! bench_rerank {
+    ($(
+        $function: ident,
+        $embedding_size: expr,
+        $document_size: expr,
+        $interest_size: expr
+    );+ $(;)?) => {$(
+        fn $function(c: &mut Criterion) {
+            let mut rng = SmallRng::from_entropy();
+            let floats = Uniform::new_inclusive(-1.0, 1.0);
+            let ints = Uniform::new(0, $interest_size);
+
+            let system = CoiConfig::default().build();
+            let timestamp = Utc::now();
+
+            let documents = (0..$document_size)
+                .map(|_| {
+                    let embedding = Embedding1::from(
+                        floats
+                            .sample_iter(&mut rng)
+                            .take($embedding_size)
+                            .collect_vec(),
+                    )
+                    .normalize()
+                    .unwrap();
+                    let tags = vec![format!("tag {}", ints.sample(&mut rng))];
+                    (embedding, tags)
+                })
+                .collect_vec();
+
+            let interests = (0..$interest_size)
+                .map(|i| {
+                    let id = CoiId::new();
+                    let point = Embedding1::from(
+                        floats
+                            .sample_iter(&mut rng)
+                            .take($embedding_size)
+                            .collect_vec(),
+                    )
+                    .normalize()
+                    .unwrap();
+                    let stats = CoiStats {
+                        view_count: i,
+                        view_time: Duration::from_secs(i as u64),
+                        last_view: timestamp,
+                    };
+                    PositiveCoi { id, point, stats }
+                })
+                .collect();
+            let interests = UserInterests {
+                positive: interests,
+                negative: Vec::new(),
+            };
+
+            let tag_weights = (0..$interest_size)
+                .map(|i| (format!("tag {i}"), i))
+                .collect::<HashMap<_, _>>();
+
+            let name = format!(
+                "rerank {} documents on {} interests (embedding size: {})",
+                $document_size,
+                $interest_size,
+                $embedding_size,
+            );
+            c.bench_function(
+                &name,
+                |b| b.iter_batched(
+                    || (documents.clone(), tag_weights.clone()),
+                    |(documents, tag_weights)| bench_rerank(
+                        black_box(&system),
+                        black_box(documents),
+                        black_box(&interests),
+                        black_box(tag_weights),
+                        black_box(timestamp),
```

Comment:
art

there is no reason to black_box the history , or any "batch" data passed with iter_batched

```suggestion
                        black_box(&system),
                        documents,
                        black_box(&interests),
                        tag_weights,
                        black_box(timestamp),
```